pallets/collator-selection/src/lib.rs

```diff
@@ -0,0 +1,434 @@
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Collator Selection pallet.
+//!
+//! A pallet to manage collators in a parachain.
+//!
+//! ## Overview
+//!
+//! The Collator Selection pallet manages the collators of a parachain. **Collation is _not_ a
+//! secure activity** and this pallet does not implement any game-theoretic mechanisms to meet BFT
+//! safety assumptions of the chosen set.
```

Comment:
> Collation is not a secure activity.

I agree with this for some definitions of "secure" but not for all. I agree that no matter how badly collators behave they can't cause the relay chain to finalize invalid parachain state transitions.

But there are still important ways in which a corrupt collator set can misbehave and cause nontrivial problems for a parachain:
1. A collator can skip slots.
2. A collator can censor transactions.

If this is couple with an author-selection mechanism that selects only one author per slot in the consensus layer (like Aura) a single malicious collator can cause the parachain to produce no block in that slot. The more collators do this the more the chain throughput decreases.

Regarding censorship, if the collator set ever becomes entirely colluding, they can retain unilateral control forever by censoring any transactions that would bond other accounts as collators.

Reply:
Actualy, I think we could construct a minimal liveness proof that a block will be authored eventually from the kick_mechanism and an assumption that the chain is launched with at least one honest collator. But the practical concern still exists. One whale could stake a lot of accounts and skip all the corresponding slots.

Adding a slash to the kick mechanism would ensure that this whale attacker would eventually bleed enough stake to not be a problem, but there would still be an unstable transient period after launch before settling down to a steady state.

Reply:
There could be a slash, but it work similarly to the relay-chain staking where the slash is non-zero only for when a threshold go offline at once (and can be superlinear afterwwards).